kobo/apps/openrosa/apps/logger/utils/attachment.py

```diff
@@ -0,0 +1,45 @@
+from django.db import transaction
+from django.db.models import F
+from django.db.models.query import QuerySet
+
+from kobo.apps.openrosa.apps.logger.models import Attachment, XForm
+from kobo.apps.openrosa.apps.main.models import UserProfile
+
+
+def bulk_update_attachment_storage_counters(
+    attachments: QuerySet[Attachment], subtract: bool
+):
+    """
+    Update storage counters in bulk based on a queryset of attachments
+
+    Args:
+        attachments: QuerySet of Attachment objects to process.
+        subtract: If True, subtract file size. If False, add file size.
+    """
+    user_storage_updates = {}
+    xform_storage_updates = {}
+    sign = -1 if subtract else 1
+
+    for attachment in attachments:
+        file_size = attachment.media_file_size
+        if not file_size:
+            continue
+
+        xform_id = attachment.instance.xform_id
+        user_id = attachment.instance.xform.user_id
+
+        xform_storage_updates.setdefault(xform_id, 0)
+        user_storage_updates.setdefault(user_id, 0)
+
+        xform_storage_updates[xform_id] += sign * file_size
+        user_storage_updates[user_id] += sign * file_size
+
+    with transaction.atomic():
+        for user_id, storage_change in user_storage_updates.items():
+            UserProfile.objects.filter(user_id=user_id).update(
+                attachment_storage_bytes=F('attachment_storage_bytes') + storage_change
+            )
+        for xform_id, storage_change in xform_storage_updates.items():
+            XForm.all_objects.filter(pk=xform_id).update(
+                attachment_storage_bytes=F('attachment_storage_bytes') + storage_change
+            )
```

Comment:
Is it possible to be more atomic and avoid calculating in Python, then updating the DB. 
Would it be possible to get the total count with PostgreSQL and the ORM?